### PR TITLE
Fix race condition in unittest by pytest temp_dir fixtures

### LINKED ISSUE #18322: test_recordimage_dataset
http://jenkins.mxnet-ci.amazon-ml.com/blue/rest/organizations/jenkins/pipelines/mxnet-validation/pipelines/windows-gpu/branches/PR-18312/runs/2/nodes/109/steps/155/log/?start=0
```

[2020-05-14T09:13:07.293Z] __________________________ test_recordimage_dataset ___________________________
[2020-05-14T09:13:07.293Z] [gw0] win32 -- Python 3.7.3 C:\Python37\python.exe
[2020-05-14T09:13:07.293Z] 
[2020-05-14T09:13:07.293Z]     @with_seed()
[2020-05-14T09:13:07.293Z]     def test_recordimage_dataset():
[2020-05-14T09:13:07.293Z]         recfile = prepare_record()
[2020-05-14T09:13:07.293Z]         fn = lambda x, y : (x, y)
[2020-05-14T09:13:07.293Z]         dataset = gluon.data.vision.ImageRecordDataset(recfile).transform(fn)
[2020-05-14T09:13:07.293Z]         loader = gluon.data.DataLoader(dataset, 1)
[2020-05-14T09:13:07.293Z]     
[2020-05-14T09:13:07.293Z] >       for i, (x, y) in enumerate(loader):
[2020-05-14T09:13:07.293Z] 
[2020-05-14T09:13:07.293Z] tests\python\unittest\test_gluon_data.py:75: 
[2020-05-14T09:13:07.293Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\gluon\data\dataloader.py:659: in same_process_iter
[2020-05-14T09:13:07.293Z]     ret = self._batchify_fn([self._dataset[idx] for idx in batch])
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\gluon\data\dataloader.py:659: in <listcomp>
[2020-05-14T09:13:07.293Z]     ret = self._batchify_fn([self._dataset[idx] for idx in batch])
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\gluon\data\dataset.py:232: in __getitem__
[2020-05-14T09:13:07.293Z]     item = self._data[idx]
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\gluon\data\vision\datasets.py:271: in __getitem__
[2020-05-14T09:13:07.293Z]     return image.imdecode(img, self._flag), header.label
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\image\image.py:211: in imdecode
[2020-05-14T09:13:07.293Z]     return cvimdecode(buf, *args, **kwargs)
[2020-05-14T09:13:07.293Z] <string>:38: in _cvimdecode
[2020-05-14T09:13:07.293Z]     ???
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\_ctypes\ndarray.py:91: in _imperative_invoke
[2020-05-14T09:13:07.293Z]     ctypes.byref(out_stypes)))
[2020-05-14T09:13:07.293Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2020-05-14T09:13:07.293Z] 
[2020-05-14T09:13:07.293Z] ret = -1
[2020-05-14T09:13:07.293Z] 
[2020-05-14T09:13:07.293Z]     def check_call(ret):
[2020-05-14T09:13:07.293Z]         """Check the return value of C API call.
[2020-05-14T09:13:07.293Z]     
[2020-05-14T09:13:07.293Z]         This function will raise an exception when an error occurs.
[2020-05-14T09:13:07.293Z]         Wrap every API call with this function.
[2020-05-14T09:13:07.293Z]     
[2020-05-14T09:13:07.293Z]         Parameters
[2020-05-14T09:13:07.293Z]         ----------
[2020-05-14T09:13:07.293Z]         ret : int
[2020-05-14T09:13:07.293Z]             return value from API calls.
[2020-05-14T09:13:07.293Z]         """
[2020-05-14T09:13:07.293Z]         if ret != 0:
[2020-05-14T09:13:07.293Z] >           raise get_last_ffi_error()
[2020-05-14T09:13:07.293Z] E           mxnet.base.MXNetError: Traceback (most recent call last):
[2020-05-14T09:13:07.293Z] E             File "..\src\io\image_io.cc", line 190
[2020-05-14T09:13:07.293Z] E           MXNetError: Check failed: len > 0: Input cannot be an empty buffer
[2020-05-14T09:13:07.293Z] 
[2020-05-14T09:13:07.293Z] windows_package\python\mxnet\base.py:246: MXNetError
[2020-05-14T09:13:07.293Z] ---------------------------- Captured stderr setup ----------------------------
```

@zhreshold

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
Fix #18322

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by leezu** [comment]
Could you clarify where the race occurs?

**Comment by zhreshold** [comment]
If multiple tests want to create the rec file at the same time, the rec file may be corrupted

**Comment by leezu** [comment]
Would creating the rec-file at a temporary path then solve the issue while preserving parallelization? You can take the `tmp_path` argument in your test function to have pytest provide a temporary directory unique for this test. https://docs.pytest.org/en/latest/tmpdir.html

**Comment by leezu** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu]

File: tests/python/unittest/test_gluon_data.py [hunk 1]
```diff
@@ -138,52 +137,52 @@ def test_sampler():
...
+    assert len(gluon.data.vision.CIFAR100(root=str(p.join('cifar100'))).__mx_handle__()) == 50000
+    assert len(gluon.data.vision.CIFAR100(root=str(p.join('cifar100')), fine_label=True).__mx_handle__()) == 50000
+    assert len(gluon.data.vision.CIFAR100(root=str(p.join('cifar100')), train=False).__mx_handle__()) == 10000
 
 @with_seed()
-def test_image_folder_dataset():
-    prepare_record()
-    dataset = gluon.data.vision.ImageFolderDataset('data/test_images')
+def test_image_folder_dataset(prepare_record):
+    print(prepare_record[:-8])
```
Comment (leezu) [inline tests/python/unittest/test_gluon_data.py:167]: https://docs.python.org/3/library/os.path.html#os.path.dirname ?

**Pull request merged**